### PR TITLE
Add support for Enums

### PR DESCRIPTION
**Pull request by irshadnilam**
## Purpose
This PR adds support for `enum` syntax. 

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/22680

## Samples
Adds support for the following
```ballerina
public enum Color {
  RED,
  GREEN,
  BLUE
}
```
is equivalent to 

```ballerina
public const RED = "RED";
public const GREEN = "GREEN";
public const BLUE = "BLUE";
public type Color RED|GREEN|BLUE;
```

Enum members can also have expressions of type string.
```ballerina
public enum Color {
  RED = "red",
  GREEN,
  BLUE
}
```
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by whizsid** [comment]
Can I use tuples with enum? This feature is already in Rust language and I like it.

```ballerina
enum {
   TWO_DIMENSIONAL(int,int);
   THREE_DIMENTIONAL(int,int,int);
}
```

**Comment by irshadnilam** [comment]
> Can I use tuples with enum? This feature is already in Rust language and I like it.
> 
> ```ballerina
> enum {
>    TWO_DIMENSIONAL(int,int);
>    THREE_DIMENTIONAL(int,int,int);
> }
> ```

@whizsid Thank you for bringing this suggestion. 

Would you able to open an issue with this suggestion on ballerina-spec repository?
https://github.com/ballerina-platform/ballerina-spec/

**Comment by hasithaa** [comment]
@irshadnilam We need more test cases, when enum value is used as function parameters, etc. This is already covered by the constants. But better to have a separate set of test cases for this.

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 1]
```diff
@@ -156,6 +157,15 @@ constantDefinition
     :   PUBLIC? CONST typeName? Identifier ASSIGN constantExpression SEMICOLON
     ;
 
+enumDefinition
+    :   documentationString? annotationAttachment* PUBLIC? ENUM Identifier LEFT_BRACE
+            (enumMember (COMMA enumMember)*)? RIGHT_BRACE
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4:162]: The grammar for module-enum-decl in the spec is 

```
module-enum-decl :=
   metadata
   [public] enum identifier { enum-member (, enum-member)* }
```
At least one enum-member is required.

So shouldn't this be

```suggestion
            enumMember (COMMA enumMember)* RIGHT_BRACE
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4:162]: The following results in a bad, sad error atm.

```ballerina
public enum Color {

}
```

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 [hunk 2]
```diff
@@ -156,6 +157,15 @@ constantDefinition
     :   PUBLIC? CONST typeName? Identifier ASSIGN constantExpression SEMICOLON
     ;
 
+enumDefinition
+    :   documentationString? annotationAttachment* PUBLIC? ENUM Identifier LEFT_BRACE
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4:161]: The spec says the attachment point for `module-enum-decl` is `type`.

But this seems to be matching for consts now.

```ballerina
public annotation X on type;
public const annotation Y on source const;

@X // should work, doesn't
@Y // should fail, doesn't
public enum Color {
    RED,
    BLUE
}
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4:161]: https://github.com/ballerina-platform/ballerina-lang/issues/23308 tracks fixing this,

File: compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4 (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/resources/grammar/BallerinaParser.g4:162]: https://github.com/ballerina-platform/ballerina-lang/issues/23308 tracks fixing this,

**Pull request merged**